V2

fedeallocati opened the pull request:
I finally finished the version 2 of the implementation:
- Protocol classes that perform the creation and parsing of packets
- Individual classes for the different kind of instructions (Ping, Write, Read, Sync Write, etc), that are templated by the Protocol
- Servo classes (one for each dynamixel model), using traits to define their Control Tables, and some handy macros to write to the memory (and also to read from the memory and parse the results), both as static and as instance methods:

```
Mx64 my_mx(1);
ctrl.send(my_mx.set_goal_position(1024));
ctrl.send(Mx64::set_goal_position(1, 1024)); // Exactly same as before

ctrl.send(Mx64::get_present_position(1));
StatusPacket<Mx64::protocol_t> status_packet;
ctrl.recv(status_packet);
std::cout << Mx64::parse_present_position(status_packet) << std::endl;

```
- A base class for all the servos that contains common functionality: Read/write of the fields that all models have (e.g. model number, id, baudrate, goal position, present speed, etc) and functionality that we build on top of the common parts (for example set_goal_position_angle, that computes the exact step for the specified angle)
- Some improvement in the error handling: Specific error clases for `CRCError` (when the dynamixel sends a package and the computed CRC doesn't match the received CRC) and `StatusError` (the error bit fields in the Status Packet). They both inherit from the base class Error.
- A small change in the `Usb2Dynamixel` controller for the timeout parameter (now it is in the class instead of the `recv` method

What is missing is the adaption of the examples and tools to the new API, and some higher level functionality like search of the available servos.

I don't expect to merge rigth now, but I need feedback on this.
CC: @dogoepp @costashatz @jbmouret

fedeallocati committed:
Initial implementation, with instructions Ping, Read and Write

fedeallocati committed:
Change in the packet-protocol-instruction interaction

fedeallocati committed:
Protocol dependant status packet

fedeallocati committed:
Protocol dependant status packet

fedeallocati committed:
Merge branch 'v2' of https://github.com/resibots/libdynamixel into v2

fedeallocati committed:
Low level static class for servos

fedeallocati committed:
Small typos and style adjustment

fedeallocati committed:
Folder structure, Pro L54, new low level instructions and some high level ones

fedeallocati committed:
Almost finished low-level instructions that existed in previous API

fedeallocati committed:
Finished API for all models of version 1

fedeallocati committed:
Control table for all models of version 1

fedeallocati committed:
Methods to parse the result from a Read instruction

fedeallocati committed:
Finished low level API

fedeallocati committed:
Added object oriented interface

fedeallocati committed:
Added XL320

fedeallocati committed:
Typed errors, first version

fedeallocati committed:
Library reestructuration

fedeallocati committed:
Small improvement to Usb2Dynamixel controller

fedeallocati committed:
Merge branch 'master' of https://github.com/resibots/libdynamixel into v2

Conflicts:
	src/dynamixel/ax12.hpp
	src/dynamixel/usb2dynamixel.cpp
	src/tools/dynamixel.cpp

dogoepp commented:
Thanks for all that work ! I'll read and comment on the code.

For now, based on your description, what about renaming the base class for exceptions to something specific to dynamixels, like DynamixelError (or put it in the dynamixel namespace) ? I find it misleading when reading code using this library and finding exceptions instances of that type. It doesn't let the programmer know where to look for the definition of the class and what it represents.

I know you kept the existing class, the original name is not your doing. :)

fedeallocati committed:
Merge branch 'master' of https://github.com/resibots/libdynamixel into v2

Conflicts:
	src/dynamixel/misc.hpp

fedeallocati committed:
Returning value instead of storing it in variable

fedeallocati committed:
Corrected bug in the integer baudrate to OS-defined baudrate function

fedeallocati committed:
Replaced boost_lexical_cast with sstream and removed boost dependency

fedeallocati commented:
> For now, based on your description, what about renaming the base class for exceptions to something specific to dynamixels, like DynamixelError (or put it in the dynamixel namespace) ? I find it misleading when reading code using this library and finding exceptions instances of that type. It doesn't let the programmer know where to look for the definition of the class and what it represents.

Actually they are inside the `dynamixel` namespace, and also the new `errors` namespace.

I will also add some comments, so it's clear how to add new models/fix current ones.

fedeallocati committed:
Small bug in Base Servo class

fedeallocati committed:
Comment on how to add new models

fedeallocati committed:
Removed old tests

fedeallocati committed:
Added a demo binary for the omnipointer arm control

fedeallocati committed:
Updated waf, wscripts and waf_tools

fedeallocati committed:
Commented arm compilation

fedeallocati committed:
Finished object oriented API

fedeallocati commented:
I finished the object oriented API.
We now have a `dynamixel::auto_detect` method (in the misc file), with inheritance and model abstraction.
With this I think that the V2 is completed and ready to merge. The arm compilation is missing(because I updated the waf version, and their API changed), and also all of the old examples, but we can add that later.

fedeallocati committed:
Added a more simple check

fedeallocati committed:
Fixed compilation in clang

fedeallocati committed:
Fixed auto_detect functionality. Added a lighter include dynamixel_core

fedeallocati committed:
Fixed gcc warning

fedeallocati committed:
Fixed a bug in include guard

dogoepp committed:
fix bug with open_serial

It used to be possible to open a new serial connexion even though the
previous one had not been closed.

dogoepp commented:
In the code, there are many assertions. Are they the proper use cases for assertions ?

cc: @jbmouret @costashatz

costashatz commented:
> In the code, there are many assertions. Are they the proper use cases for assertions ?
> 
> cc: @jbmouret @costashatz

I will have a look this week... I am not convinced yet that all these assertions should be there.

fedeallocati commented:
They should'nt be, but it was the quickest way to leave a TODO for throwing
exceptions :P

2016-03-01 10:58 GMT+01:00 Konstantinos Chatzilygeroudis <
notifications@github.com>:

> In the code, there are many assertions. Are they the proper use cases for
> assertions ?
> 
> cc: @jbmouret https://github.com/jbmouret @costashatz
> https://github.com/costashatz
> 
> I will have a look this week... I am not convinced yet that all these
> assertions should be there.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/resibots/libdynamixel/pull/6#issuecomment-190641600.

dogoepp committed:
Merge pull request #7 from resibots/v2-dorian

fix bug with open_serial

dogoepp committed:
Explaining a bit unpack_status and StatusPacket

dogoepp committed:
remove timeout parameter from open_serial

The method open_serial could be called several times in the lifecycle of
the class but we might not want to change the recieve timeout, nor reset
it to the default value. Therefore, it is initialised by the constructor
(isn't it cleaner anyway ?) and still accessible with a getter and a
setter.

dogoepp committed:
Add default value to empty constructor

dogoepp commented:
It seems that we love typedefs. Then, why don't we use one for the packet's parameters (like [here](https://github.com/resibots/libdynamixel/pull/6/files#diff-6060f3580334ede77bbf3c80f6b8478aR53)) ? We already use typedefs for `instr_t`, even though it maps to the same type for protocol 1 and 2.

dogoepp committed:
add spaces

dogoepp committed:
comments for unpack_status

dogoepp committed:
fix error handling in protocol 2

dogoepp commented:
What is the difference between a reg_goal_position_angle and a set_goal_position_angle ?

dogoepp commented on src/dynamixel/servos/servo.hpp:
```diff
@@ -0,0 +1,243 @@
+    \
+    inline InstructionPacket<typename Servo<Model>::protocol_t> set_##Name(typename Servo<Model>::ct_t::Name##_t value)  const \
+    { \
+            return Model::set_##Name(this->_id, value); \
```
Which method is this calling ?

fedeallocati replied:
I'm not sure that I understand your question.
We always have 2 versions of a method:
The static version, accesed trough the model, e.g. `Mx28::set_moving_speed(id, speed)`,
and the instance version, which calls the static version of the same function, but passing the stored id.

dogoepp replied:
You answered the question orally, thanks.

dogoepp committed:
whitespaces

fedeallocati commented:
Reg methods store the value in a temporary memory of the servos, that gets
written to the RAM or ROM when you send it an Action instruction.
That's what we are using in dynamixel_control_hw to send the joint
positions all at the same time: we first register the positions in all the
servos, and then broadcast an Action instruction
El mar. 2, 2016 8:59, "Dorian Goepp" notifications@github.com escribió:

> What is the difference between a reg_goal_position_angle and a
> set_goal_position_angle ?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/resibots/libdynamixel/pull/6#issuecomment-191118180.

dogoepp commented:
OK thanks

dogoepp committed:
Explain a bit auto_detect

dogoepp commented:
As soon as all previous comments are solved, I'm ok for merging.

fedeallocati committed:
Changed from typename T to typename Protocol

fedeallocati committed:
Changed from typename T to typename Protocol

dogoepp committed:
fix binary and

dogoepp committed:
put all error codes in hexadecimal

fedeallocati committed:
Changing Servo constructor access right to protected

dogoepp committed:
put reference to the projets using this library

dogoepp commented:
@costashatz Did you have a look at the asserts ? Shall I do this ?

costashatz commented:
@dogoepp I will do it this weekend.  I would have done it today but I got
held up by a bug in hexapod_dart..

On Fri, Mar 4, 2016, 17:38 Dorian Goepp notifications@github.com wrote:

> @costashatz https://github.com/costashatz Did you have a look at the
> asserts ? Shall I do this ?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/resibots/libdynamixel/pull/6#issuecomment-192346821.
> 
> ## 
> 
> Konstantinos Chatzilygeroudis
> PhD Candidate
> Team Larsen
> INRIA Nancy Grand-Est
> Room C121
> e-mail: konstantinos.chatzilygeroudis@inria.fr
> web: http://costashatz.github.io/

costashatz committed:
First try for assertions fix

costashatz committed:
Merge remote-tracking branch 'origin/v2' into assert_fix

Conflicts:
	src/dynamixel/protocols/protocol1.hpp

costashatz committed:
More exceptions instead of assertions

costashatz commented:
> @costashatz Did you have a look at the asserts ? Shall I do this ?

@dogoepp @fedeallocati please look at #11 ..

dogoepp committed:
Improvement proposal for error messages.

dogoepp committed:
Merge pull request #12 from resibots/assert_fix_messages

Proposal for improvement of error messages

costashatz committed:
Merge pull request #11 from resibots/assert_fix

Assertions fix

dogoepp committed:
finish the switch to hexadecimal values

dogoepp commented:
Please have a look at #13.

Besides, ready to merge ?

costashatz commented:
> Please have a look at #13.
> 
> Besides, ready to merge ?

Once we merge #13, I will format the code. Then we are ready to merge..

dogoepp committed:
Merge pull request #13 from resibots/v2-errors

Fixing handling of errors in Protocol v2

dogoepp commented:
When do you think you coud take care of the formating @costashatz ? Do you need help on this ?

dogoepp committed:
apply clang-format

dogoepp commented:
OK to merge ?

costashatz commented:
> OK to merge ?

There are conflicts...

dogoepp commented:
Because I updated v1 but the related file is deleted in v2 (utility 
program).

Le 11/03/2016 16:44, Konstantinos Chatzilygeroudis a écrit :

> ```
> OK to merge ?
> ```
> 
> There are conflicts...
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/resibots/libdynamixel/pull/6#issuecomment-195421710.

costashatz commented:
> Because I updated v1 but the related file is deleted in v2 (utility
> program).

Yes but we need to fix it..

dogoepp committed:
Apply clang-format, this time with our style
